alua.c

```diff
@@ -560,17 +561,49 @@ int alua_implicit_transition(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 		return ret;
 
 	pthread_mutex_lock(&rdev->state_lock);
-	if (rdev->lock_state == TCMUR_DEV_LOCK_LOCKED) {
+	if (rdev->lock_state == TCMUR_DEV_LOCK_WRITE_LOCKED) {
+		/* For both read/write cases in this state is good */
 		goto done;
-	} else if (rdev->lock_state == TCMUR_DEV_LOCK_LOCKING) {
-		tcmu_dev_dbg(dev, "Lock acquisition operation is already in process.\n");
+	} else if (rdev->lock_state == TCMUR_DEV_LOCK_WRITE_LOCKING) {
+		/* For both read/write cases in this state should return busy */
+		tcmu_dev_dbg(dev, "Write lock acquisition operation is already in process.\n");
 		ret = TCMU_STS_BUSY;
 		goto done;
+	} else if (is_read && !rdev->lock_lost) {
+		if (rdev->lock_state == TCMUR_DEV_LOCK_READ_LOCKED) {
+			goto done;
+		} else if (rdev->lock_state == TCMUR_DEV_LOCK_READ_LOCKING) {
+			tcmu_dev_dbg(dev, "Read lock acquisition operation is already in process.\n");
+			ret = TCMU_STS_BUSY;
+			goto done;
+		}
 	}
 
-	tcmu_dev_info(dev, "Starting lock acquisition operation.\n");
+	if (is_read) {
+		tcmu_dev_info(dev, "Starting read lock acquisition operation.\n");
+
+		/*
+		 * Since we are here, current lock state should be one of:
+		 *   TCMUR_DEV_LOCK_UNLOCKED
+		 *   TCMUR_DEV_LOCK_UNKNOWN
+		 *
+		 * Will not acquire the lock, just reopen the device.
+		 */
+		rdev->lock_state = TCMUR_DEV_LOCK_READ_LOCKING;
+	} else {
+		tcmu_dev_info(dev, "Starting write lock acquisition operation.\n");
 
-	rdev->lock_state = TCMUR_DEV_LOCK_LOCKING;
+		/*
+		 * Since we are here, current lock state should be one of:
+		 *   TCMUR_DEV_LOCK_UNLOCKED
+		 *   TCMUR_DEV_LOCK_UNKNOWN
+		 *   TCMUR_DEV_LOCK_READ_LOCKING
+		 *   TCMUR_DEV_LOCK_READ_LOCKED
+		 *
+		 * May will reopen the deivce and Will acquire the lock later.
+		 */
+		rdev->lock_state = TCMUR_DEV_LOCK_WRITE_LOCKING;
```

Comment:
If you are in `READ_LOCKING` state so it's attempting to open the image and now a write comes in, wouldn't this cause a second thread to start re-opening the image and attempting to acquire the write lock? Should it instead just return `BUSY` in this case to retry acquiring the write lock -- especially because `tcmu_acquire_dev_lock` acquires and drops the lock numerous times so it seems like lots of potential for race conditions.

I know you said it will be a busy loop between the kernel and tcmu-runner when the `STS_BUSY` result is handled, but it's going to be doing that regardless on successive reads while re-opening the image and on successive writes.

Reply:
> If you are in `READ_LOCKING` state so it's attempting to open the image and now a write comes in, wouldn't this cause a second thread to start re-opening the image and attempting to acquire the write lock?

It's a good question but the re-opening issue shouldn't happen, but the acquiring the write lock in parallel may will:

There has 3 cases:
Case1:
If the first thread still not finished from `tcmu_reopen_dev()`, it will set `rdev->flags |= TCMUR_DEV_FLAG_IN_RECOVERY`, so for the second thread, though it may set `reopen = true`, because the first thread may still not set the `rdev->lock_lost = false`. The second thread will just get a `EBUSY` error from `tcmu_reopen_dev()`, then `goto drop_conn` without trying to acquire the lock. But the second have already upgrade the lock state to `TCMUR_DEV_LOCK_WRITE_LOCKING`. So the first thread will try to acquire the lock after it returned from `tcmu_reopen_dev()`.

Case2:
If the first thread has just returned from `tcmu_reopen_dev()` when the second thread begins to run, because the first thread has already set the `rdev->lock_lost = false` in `tcmu_reopen_dev()`, the it won't do the device reopen. But it is possible that both the two threads will see that the lock state is `TCMUR_DEV_LOCK_WRITE_LOCKING` and then do the `rhandler->lock()` in parallel. I think I need to fix this.

The above two cases, if just return `EBUSY` here, it works well. But for the following Case3, it won't. 

Case3:
If the first thread has already finished from `tcmu_reopen_dev()` and is runing the code after `drop_conn:`, that means the first thread didn't see the lock state is `TCMUR_DEV_LOCK_WRITE_LOCKING`, so it didn't acquire the lock. And then we need the second thread to continue to acquire lock only.  This is why I add the `no_lock` in `tcmu_acquire_dev_lock()`.


> Should it instead just return `BUSY` in this case to retry acquiring the write lock -- especially because `tcmu_acquire_dev_lock` acquires and drops the lock numerous times so it seems like lots of potential for race conditions.
> 
> I know you said it will be a busy loop between the kernel and tcmu-runner when the `STS_BUSY` result is handled, but it's going to be doing that regardless on successive reads while re-opening the image and on successive writes.

Yeah, the retry happens in the initiator's scsi driver in kernel space. This logic should be in compliance with the scsi proto.

Reply:
Right, it just seems cleaner to let it retry via `STS_BUSY` instead of kicking off multiple threads.

Reply:
Yeah, this sounds cool. And I have fixed it.